pubsub-mapped-api/src/main/java/com/google/pubsub/clients/consumer/KafkaConsumer.java

```diff
@@ -0,0 +1,592 @@
+      Deserializer<K> keyDeserializer,
+      Deserializer<V> valueDeserializer) {
+    this(new ConsumerConfig(
+            ConsumerConfig.addDeserializerToConfig(properties, keyDeserializer, valueDeserializer)),
+        keyDeserializer, valueDeserializer);
+  }
+
+
+  private KafkaConsumer(ConsumerConfig configs, Deserializer<K> keyDeserializer,
+      Deserializer<V> valueDeserializer) {
+    this(configs, keyDeserializer, valueDeserializer, new StubCreator(ChannelUtil.getInstance()));
+  }
+
+  @SuppressWarnings("unchecked")
+  KafkaConsumer(ConsumerConfig configs, Deserializer<K> keyDeserializer,
+      Deserializer<V> valueDeserializer, StubCreator stubCreator) {
+    try {
+      log.debug("Starting PubSub subscriber");
+
+      this.stubCreator = stubCreator;
+      this.keyDeserializer = handleDeserializer(configs,
+          ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, keyDeserializer, true);
+      this.valueDeserializer = handleDeserializer(configs,
+          ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, valueDeserializer, false);
+
+      //this is a limit on each poll for each topic
+      this.maxPollRecords = configs.getInt(ConsumerConfig.MAX_POLL_RECORDS_CONFIG);
+
+      log.debug("PubSub subscriber created");
+    } catch (Throwable t) {
+      throw new KafkaException("Failed to construct PubSub subscriber", t);
+    }
+  }
+
+  private Deserializer handleDeserializer(ConsumerConfig configs,
+      String configString, Deserializer providedDeserializer, boolean isKey) {
+    Deserializer deserializer;
+    if(providedDeserializer == null) {
+      deserializer = configs.getConfiguredInstance(configString, Deserializer.class);
+      deserializer.configure(configs.originals(), isKey);
+    } else {
+      configs.ignore(configString);
+      deserializer = providedDeserializer;
+    }
+    return deserializer;
+  }
+
+  public Set<TopicPartition> assignment() {
+    throw new NotImplementedException("Not yet implemented");
+  }
+
+  public Set<String> subscription() {
+    return subscriptions.keySet();
+  }
+
+  /*
+  For every new consumer, a temporary subscription is created. After "unsubscribe" operation
+  temporary subscriptions are removed. After an error while creating subscriptions the attempt
+  is made to delete all temporary subscriptions made in this call to "subscribe".
+
+  Every version of subscribe is replacing old subscriptions with new ones.
+
+  Temporary subscription names are built as a combination of timestamp and randomly generated UUID.
+   */
+  public void subscribe(Collection<String> topics, ConsumerRebalanceListener listener) {
+    checkSubscribePreconditions(topics);
+
+    unsubscribe();
+    List<ResponseData<Subscription>> futureSubscriptions = deputePubsubSubscribes(topics);
+    Map<String, Subscription> subscriptionMap = getNewSubscriptionMap(futureSubscriptions);
+
+    log.debug("Subscribed to topic(s): {}", Utils.join(subscriptionMap.keySet(), ", "));
```

Comment:
this is exactly as in kafka code. i thought that means that evaluation is lazy in such situation, isn't it?